Reorganize application files and refactor gateways

aabadie opened the pull request:
This PR fixes several inconsistencies and bad naming in all applications (broker, dashboard, gateways):
* `application.py` and specific application modules (gateway.py, broker.py, etc) have been exchanged, making the module names more consistent
* The gateway, broker and dashboard classes have also been renamed

The last big change removes code duplication in gateways by introducing a base gateway object.

aabadie commented:
Maybe @bergzand can give some help to review the changes here.

bergzand commented:
That's a lot of refactoring! I should have some time this weekend for a review.

bergzand commented on pyaiot/gateway/common/common.py:
```diff
@@ -0,0 +1,116 @@
+
+
+class GatewayBase(web.Application):
+    """Tornado based gateway application for CoAP nodes on a network."""
```
Maybe reword to "Base class for Tornado based gateway applications" as this is not a CoAP specific base class

bergzand commented on pyaiot/gateway/common/common.py:
```diff
@@ -0,0 +1,116 @@
+
+
+    def setup_nodes_controller(self):
+        """Instanciate and configure an application nodes controller."""
```
Small typo: "Instantiate"

bergzand commented on pyaiot/gateway/mqtt/gateway.py:
```diff
@@ -27,81 +27,34 @@
-        logger.error("Config file not found: {}".format(e))
-        return
+    def setup_nodes_controller(self):
+        """Instanciate and configure an MQTT nodes controller."""
```
Small typo: "Instantiate"

bergzand commented on pyaiot/gateway/common/common.py:
```diff
@@ -0,0 +1,116 @@
+        settings = {'debug': True}
+
+        # Setup nodes controller
+        self._nodes_controller = self.setup_nodes_controller()
```
Not completely sure on this, but maybe add a dummy `setup_nodes_controller()` to this class

aabadie replied:
You hit the weak point of this PR :)
With this version, it can happen that a concrete gateway doesn't implement this method and the `_nodes_controller` could be `None`

I changed that to use multiple inheritance using `ABCMeta` and setting the `setup_nodes_controller` as pure virtual. Then all concrete gateway implementations have to implement this method. I tested it and it works well: an error message is displayed when the method is not implemented.
We could go even further by applying a similar design to the node controllers: a base class with abstract methods and concrete classes implementing them.

bergzand replied:
Nice, this is exactly what I hoped to achieve with this comment.

bergzand commented on pyaiot/gateway/coap/gateway.py:
```diff
@@ -27,80 +27,31 @@
+        logger.info('CoAP gateway application started')
 
+    def setup_nodes_controller(self):
+        """Instanciate and configure a CoAP nodes controller."""
```
Small typo: "Instantiate"

bergzand commented on pyaiot/gateway/ws/gateway.py:
```diff
@@ -27,73 +27,112 @@
+                    .format(options.gateway_port))
+
+    def setup_nodes_controller(self):
+        """Instanciate and configure a websocket nodes controller."""
```
Small typo: "Instantiate"

bergzand commented:
Looks good overall, I didn't test it yet, but the overall changes look good to me. Good work with the refactor to a single base gateway class 👍

bergzand commented:
There are also a few pep8 issues. Do you want to enforce pep8 on PR changes?

aabadie commented:
@bergzand, Thanks for reviewing this. I addressed your comments.

> There are also a few pep8 issues. Do you want to enforce pep8 on PR changes?

Yes, it's always a good thing ! This can be done very easily with pytest. Do you propose yourself for a PR ?

bergzand commented:
> Yes, it's always a good thing ! This can be done very easily with pytest. Do you propose yourself for a PR ?

Yes, I can get a PR ready with pep8 fixes and an additional test this week or so.

Tested a bit, looks good.
Ack from me for this PR

aabadie commented:
> Ack from me for this PR

Thanks, then I'll merge !
